src/extractor/edge_based_graph_factory.cpp

```diff
@@ -368,181 +372,267 @@ void EdgeBasedGraphFactory::GenerateEdgeExpandedEdges(
         util::Percent progress(log, m_node_based_graph->GetNumberOfNodes());
         // going over all nodes (which form the center of an intersection), we compute all
         // possible turns along these intersections.
-        for (const auto node_at_center_of_intersection :
-             util::irange(0u, m_node_based_graph->GetNumberOfNodes()))
-        {
-            progress.PrintStatus(node_at_center_of_intersection);
-
-            const auto shape_result =
-                turn_analysis.ComputeIntersectionShapes(node_at_center_of_intersection);
-
-            // all nodes in the graph are connected in both directions. We check all outgoing nodes
-            // to
-            // find the incoming edge. This is a larger search overhead, but the cost we need to pay
-            // to
-            // generate edges here is worth the additional search overhead.
-            //
-            // a -> b <-> c
-            //      |
-            //      v
-            //      d
-            //
-            // will have:
-            // a: b,rev=0
-            // b: a,rev=1 c,rev=0 d,rev=0
-            // c: b,rev=0
-            //
-            // From the flags alone, we cannot determine which nodes are connected to `b` by an
-            // outgoing
-            // edge. Therefore, we have to search all connected edges for edges entering `b`
-            for (const EdgeID outgoing_edge :
-                 m_node_based_graph->GetAdjacentEdgeRange(node_at_center_of_intersection))
-            {
-                const NodeID node_along_road_entering =
-                    m_node_based_graph->GetTarget(outgoing_edge);
-
-                const auto incoming_edge = m_node_based_graph->FindEdge(
-                    node_along_road_entering, node_at_center_of_intersection);
-
-                if (m_node_based_graph->GetEdgeData(incoming_edge).reversed)
-                    continue;
+        tbb::mutex merge_mutex; // To control merging of all the data into the main data structures
+        tbb::mutex bearing_hash_mutex; // To control entry into the two hash maps
+        tbb::mutex entry_hash_mutex;   // To control entry into the two hash maps
+        tbb::mutex turn_lane_mutex;    // To control entry into the two hash maps
+        tbb::concurrent_vector<BearingClassID> threadsafe_bearing_class_by_node_based_node;
+
+        threadsafe_bearing_class_by_node_based_node.resize(
+            bearing_class_by_node_based_node.size(), std::numeric_limits<std::uint32_t>::max());
+
+        tbb::parallel_for(
+            tbb::blocked_range<NodeID>(0, m_node_based_graph->GetNumberOfNodes()),
+            [&](const tbb::blocked_range<NodeID> &intersection_node_range) {
+
+                // We capture the thread-local work in these objects, then flush
```

Comment:
TBB has a nice pattern for the thread-local accumulation and serial merge: [`tbb::enumerable_thread_local`](https://software.intel.com/en-us/node/506244).

The `vector`s would be moved outside of the `parallel_for` loop as `tbb::enumerable_thread_specific` and instead of having a mutex in the `parallel_for` loop we could simply use a serial for loop after this block to iterate over all thread-local buffer and merge them.